### PR TITLE
Fixing inconsistent IRAM_ATTR definitions across compilation units (IDFGH-9403)

### PR DESCRIPTION
**Pull request by quiret**
The IRAM_ATTR preprocessor definitions depend on the GCC `__COUNTER__` preprocessor macro that automatically increments itself after being used in code. Using this feature is a bad design decision inside of headers since they can be included prior to other `__COUNTER__` uses, making the definitions that use `__COUNTER__` different across compilation units (only fixed in the C++20 modules design). This is causing confusing compiler warnings in latest GCC 11.2.0 from Xtensa and I guess many developers added the -Wno-attributes flag to their building process because of it.

This pull request fixed the issue. All of the depending projects (Arduino Core for example) would have those annoying build warnings. And, oh boy, is our world connected. I suggest you to **backport this fix** (4.4.3).

> C:/Users/digit/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32/include/freertos/port/xtensa/include/freertos/portmacro.h:613:48: warning: ignoring attribute 'section (".iram1.24")' because it conflicts with previous 'section (".iram1.22")' [-Wattributes]
>   613 | static inline bool IRAM_ATTR xPortCanYield(void)

### CODE REVIEW

**Comment by Alvin1Zhang** [comment]
Thanks for your contribution.

**Comment by igrr** [comment]
I think the solution we would like to have is to remove IRAM_ATTR from function declarations in the header file, except for the static inline definitions (which don't have this attribute mismatch issue). Would that work for you?

**Comment by igrr** [comment]
Wait, i have just realized that you are trying to compile IDF v4.4.3 (via the Arduino project?) with GCC 11.2. That's probably not going to work, as IDF v4.4.x is not compatible with the new GCC 11 based toolchain. We have already fixed all the attribute related warnings in IDF v5.0, which does support the GCC 11 based toolchain.

**Comment by quiret** [comment]
> Wait, i have just realized that you are trying to compile IDF v4.4.3 (via the Arduino project?) with GCC 11.2. That's probably not going to work, as IDF v4.4.x is not compatible with the new GCC 11 based toolchain. We have already fixed all the attribute related warnings in IDF v5.0, which does support the GCC 11 based toolchain.

How has IRAM_ATTR been fixed if `__COUNTER__` is still being used?

https://github.com/espressif/esp-idf/blob/release/v5.0/components/esp_common/include/esp_attr.h

Do you really think that merely removing the functions now will fix this broken design decision? Are you sure that future developers will not fall into the same pitfall, making the warning reappear? Are you sure devs won't be ignorant by simply using -Wno-attributes?

**Comment by igrr** [comment]
You are completely right that removing IRAM_ATTR from the function declarations doesn't prevent future developers from accidentally adding it back again. For ESP-IDF and other projects maintained by Espressif, we make sure that such issues will be discovered by running CI builds with additional warnings-as-errors flags enabled.

Of course, a developer of another project might encounter a -Wattributes warning and decide to disable this warning instead of fixing it. (Which may happen for any other useful compiler warning, right?)

> Also you mentioning this word "trying" does not 100% reflect the reality.

I apologize for misrepresenting what you are doing! I didn't realize this when I made the comment, now I have read the linked PR and I think I understand what you are doing. Indeed, if ESP-IDF v4.4.x is already pre-compiled (via the arduino-esp32 project) then building other code with GCC 11 and linking it to these pre-compiled libraries will work fine.

On the topic of the PR itself, unfortunately compiling a "blink" example with your PR branch currently results in several errors:

```
In file included from /Users/ivan/e/esp-idf/components/hal/include/hal/adc_types.h:13,
                 from /Users/ivan/e/esp-idf/components/ulp/ulp_common/include/ulp_adc.h:9,
                 from /Users/ivan/e/esp-idf/components/ulp/ulp_common/ulp_adc.c:8:
/Users/ivan/e/esp-idf/components/esp_common/include/esp_attr.h:189: warning: "EXT_RAM_ATTR" redefined
  189 | #define EXT_RAM_ATTR _Pragma ("GCC warning \"'EXT_RAM_ATTR' macro is deprecated, please use `EXT_RAM_BSS_ATTR`\"")
      | 
In file included from /Users/ivan/e/esp-idf/components/hal/include/hal/adc_types.h:13,
                 from /Users/ivan/e/esp-idf/components/ulp/ulp_common/include/ulp_adc.h:9,
                 from /Users/ivan/e/esp-idf/components/ulp/ulp_common/ulp_adc.c:8:
/Users/ivan/e/esp-idf/components/esp_common/include/esp_attr.h:94: note: this is the location of the previous definition
   94 | #define EXT_RAM_ATTR
      | 
```

After fixing the duplicate definitions in esp_attr.h, I get:
```
/Users/ivan/e/esp-idf/components/spi_flash/spi_flash_chip_generic.c:51:41: error: 'esp_flash_encryption_default' causes a section type conflict with 'spi_flash_chip_generic_timeout'
   51 | 
    static spi_flash_encryption_t esp_flash_encryption_default __attribute__((__unused__)) = {
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ivan/e/esp-idf/components/spi_flash/spi_flash_chip_generic.c:72:41: note: 'spi_flash_chip_generic_timeout' was declared here
   72 | const DRAM_ATTR flash_chip_op_timeout_t spi_flash_chip_generic_timeout = {
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

To debug this, I have ran just the preprocessor on this file. The relevant part is:
```
__attribute__((section(".dram1" "." "__ESP_COUNTER__"))) static spi_flash_encryption_t esp_flash_encryption_default __attribute__((__unused__)) = {
    .flash_encryption_enable = spi_flash_encryption_hal_enable,
    .flash_encryption_disable = spi_flash_encryption_hal_disable,
    .flash_encryption_data_prepare = spi_flash_encryption_hal_prepare,
    .flash_encryption_done = spi_flash_encryption_hal_done,
    .flash_encryption_destroy = spi_flash_encryption_hal_destroy,
    .flash_encryption_check = spi_flash_encryption_hal_check,
};
```

Basically it looks like the section name gets expanded to `.dram1.__ESP_COUNTER__`, not to the actual counter value. Could you please check this? If you get a chance to fix this, please ping me and I'll take a look at the PR again.

**Comment by adamczapski** [comment]
If I'm not mistaken, one way to solve this without changing `IRAM_ATTR` and other definitions would be to use them only in function definitions (i.e. in *.c files). In the declarations in header files you would have to remove them and add `extern` keyword instead.

For example, `cache_utils.c` remains the same, i.e.:
```
...
void IRAM_ATTR esp32_spiflash_start(void)
{
    ....
```
But in `cache_utils.h` it changes to:
```
...
extern void esp32_spiflash_start(void);
...
```

**Comment by KonstantinKondrashov** [comment]
I close this PR for now because it does not fit IDF, it makes several errors compiling a "blink" example (https://github.com/espressif/esp-idf/pull/10773#issuecomment-1434343718). Feel free to update it or open a new one.
Thanks.

**Comment by murarduino** [comment]
> You are completely right that removing IRAM_ATTR from the function declarations doesn't prevent future developers from accidentally adding it back again. For ESP-IDF and other projects maintained by Espressif, we make sure that such issues will be discovered by running CI builds with additional warnings-as-errors flags enabled.
> 
> Of course, a developer of another project might encounter a -Wattributes warning and decide to disable this warning instead of fixing it. (Which may happen for any other useful compiler warning, right?)
> 
> > Also you mentioning this word "trying" does not 100% reflect the reality.
> 
> I apologize for misrepresenting what you are doing! I didn't realize this when I made the comment, now I have read the linked PR and I think I understand what you are doing. Indeed, if ESP-IDF v4.4.x is already pre-compiled (via the arduino-esp32 project) then building other code with GCC 11 and linking it to these pre-compiled libraries will work fine.
> 
> On the topic of the PR itself, unfortunately compiling a "blink" example with your PR branch currently results in several errors:
> 
> ```
> In file included from /Users/ivan/e/esp-idf/components/hal/include/hal/adc_types.h:13,
>                  from /Users/ivan/e/esp-idf/components/ulp/ulp_common/include/ulp_adc.h:9,
>                  from /Users/ivan/e/esp-idf/components/ulp/ulp_common/ulp_adc.c:8:
> /Users/ivan/e/esp-idf/components/esp_common/include/esp_attr.h:189: warning: "EXT_RAM_ATTR" redefined
>   189 | #define EXT_RAM_ATTR _Pragma ("GCC warning \"'EXT_RAM_ATTR' macro is deprecated, please use `EXT_RAM_BSS_ATTR`\"")
>       | 
> In file included from /Users/ivan/e/esp-idf/components/hal/include/hal/adc_types.h:13,
>                  from /Users/ivan/e/esp-idf/components/ulp/ulp_common/include/ulp_adc.h:9,
>                  from /Users/ivan/e/esp-idf/components/ulp/ulp_common/ulp_adc.c:8:
> /Users/ivan/e/esp-idf/components/esp_common/include/esp_attr.h:94: note: this is the location of the previous definition
>    94 | #define EXT_RAM_ATTR
>       | 
> ```
> 
> After fixing the duplicate definitions in esp_attr.h, I get:
> 
> ```
> /Users/ivan/e/esp-idf/components/spi_flash/spi_flash_chip_generic.c:51:41: error: 'esp_flash_encryption_default' causes a section type conflict with 'spi_flash_chip_generic_timeout'
>    51 | 
>     static spi_flash_encryption_t esp_flash_encryption_default __attribute__((__unused__)) = {
>       |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /Users/ivan/e/esp-idf/components/spi_flash/spi_flash_chip_generic.c:72:41: note: 'spi_flash_chip_generic_timeout' was declared here
>    72 | const DRAM_ATTR flash_chip_op_timeout_t spi_flash_chip_generic_timeout = {
>       |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> To debug this, I have ran just the preprocessor on this file. The relevant part is:
> 
> ```
> __attribute__((section(".dram1" "." "__ESP_COUNTER__"))) static spi_flash_encryption_t esp_flash_encryption_default __attribute__((__unused__)) = {
>     .flash_encryption_enable = spi_flash_encryption_hal_enable,
>     .flash_encryption_disable = spi_flash_encryption_hal_disable,
>     .flash_encryption_data_prepare = spi_flash_encryption_hal_prepare,
>     .flash_encryption_done = spi_flash_encryption_hal_done,
>     .flash_encryption_destroy = spi_flash_encryption_hal_destroy,
>     .flash_encryption_check = spi_flash_encryption_hal_check,
> };
> ```
> 
> Basically it looks like the section name gets expanded to `.dram1.__ESP_COUNTER__`, not to the actual counter value. Could you please check this? If you get a chance to fix this, please ping me and I'll take a look at the PR again.

Does removing "IRAM_ATTR" mean that the function can only run in Flash and not RAM?
